openlr/paths_connector.cpp

```diff
@@ -0,0 +1,220 @@
+
+using namespace std;
+
+namespace openlr
+{
+namespace
+{
+// @returns a length of the longest suffix of |a| than matches any prefix of |b|.
+inline uint32_t PathOverlappingLen(Graph::EdgeVector const & a, Graph::EdgeVector const & b)
+{
+  auto aIt = find(begin(a), end(a), b.front());
+  if (aIt == end(a))
+    return 0;
+  uint32_t length = 0;
+  for (auto bIt = begin(b); aIt != end(a) && bIt != end(b);)
+  {
+    if (EdgesAreAlmostEqual(*aIt++, *bIt++))
+      ++length;
+    else
+      return 0;
+  }
+  return length;
+}
+}  // namespace
+
+bool PathsConnector::ConnectCandidates(vector<LocationReferencePoint> const & points,
+                                       vector<vector<Graph::EdgeVector>> const & lineCandidates,
+                                       vector<Graph::EdgeVector> & resultPath)
+{
+  resultPath.resize(points.size() - 1);
+
+  // TODO(mgsergio): Discard last point on failure.
+  // TODO(mgsergio): Do not iterate more than kMaxRetries times.
+  // TODO(mgserigio): Make kMaxRetries depend on points number in the segment.
+  for (size_t i = 1; i < points.size(); ++i)
+  {
+    bool found = false;
+
+    auto const distanceToNextPoint = points[i - 1].m_distanceToNextPoint;
+    auto const & fromCandidates = lineCandidates[i - 1];
+    auto const & toCandidates = lineCandidates[i];
+
+    Graph::EdgeVector fakePath;
+
+    for (size_t fromInd = 0; fromInd < fromCandidates.size() && !found; ++fromInd)
+    {
+      for (size_t toInd = 0; toInd < toCandidates.size() && !found; ++toInd)
+      {
+        resultPath[i - 1].clear();
+
+        found = ConnectAdjacentCandidateLines(points[i - 1].m_functionalRoadClass,
+                                              distanceToNextPoint, fromCandidates[fromInd],
+                                              toCandidates[toInd], resultPath[i - 1]);
+
+        if (!found)
+          continue;
+
+        found = ValidatePath(resultPath[i - 1], distanceToNextPoint, m_pathLengthTolerance);
+        if (fakePath.empty() && found &&
+            (resultPath[i - 1].front().IsFake() || resultPath[i - 1].back().IsFake()))
+        {
+          fakePath = resultPath[i - 1];
+          found = false;
+        }
+      }
+    }
+
+    if (!fakePath.empty() && !found)
+    {
+      found = true;
+      resultPath[i - 1] = fakePath;
+    }
+
+    if (!found)
+    {
+      LOG(LDEBUG, ("No shortest path found"));
+      ++m_stat.m_noShortestPathFound;
+      resultPath[i - 1].clear();
+      return false;
+    }
+  }
+
+  ASSERT_EQUAL(resultPath.size(), points.size() - 1, ());
+
+  return true;
+}
+
+bool PathsConnector::FindShortestPath(FunctionalRoadClass const frc, uint32_t const maxPathLength,
+                                      Graph::Edge const & from, Graph::Edge const & to,
+                                      Graph::EdgeVector & path)
+{
+  // TODO(mgsergio): Turn dijkstra to A*.
+
+  ScopedTimer t(m_stat.m_shortestPathsTime);
+
+  uint32_t const kLengthToleranceM = 10;
```

Comment:
Please, move this const to the top of the method.

Reply:
Done.